pkg/upstream/topo.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package upstream
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb-dashboard/util/distro"
+	"github.com/pingcap/tidb-dashboard/util/netutil"
+	"github.com/pingcap/tidb/pkg/domain/infosync"
+	"github.com/pingcap/tiflow/pkg/errors"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const (
+	// topologyTiCDC is /topology/ticdc/{clusterID}/{ip:port}.
+	topologyTiCDC = "/topology/ticdc/%s/%s"
+	// topologyTiDB is /topology/tidb/{ip:port}.
+	// Refer to https://github.com/pingcap/tidb/blob/release-7.5/pkg/domain/infosync/info.go#L78-L79.
+	topologyTiDB    = infosync.TopologyInformationPath
+	topologyTiDBTTL = infosync.TopologySessionTTL
+	// defaultTimeout is the default timeout for etcd and mysql operations.
+	defaultTimeout = time.Second
+)
+
+type tidbInstance struct {
+	IP   string
+	Port uint
+}
+
+func fetchTiDBTopology(ctx context.Context, etcdClient *clientv3.Client) ([]tidbInstance, error) {
```

Comment:
Could you please add some comments to describe what the function works for clearly before these functions? It will makes readers get the point more easier.